### PR TITLE
Use EnumSerializer for explicitly serializable enum instead of auto-generated

### PR DESCRIPTION
**Pull request by shanshin**
Resolves Kotlin/kotlinx.serialization#683 Kotlin/kotlinx.serialization#1372

### CODE REVIEW

**Comment by qwwdfsad** [comment]
Is it going to 1.7.0 or 1.7.20?

**Comment by shanshin** [comment]
> Is it going to 1.7.0 or 1.7.20?

`1.7.20`

**Comment by shanshin** [comment]
Merge after switching kotlinx.serialization to the kotlin  version `1.7.20`

File: core/commonMain/src/kotlinx/serialization/internal/Enums.kt [hunk 1]
```diff
@@ -49,20 +50,54 @@ internal class EnumDescriptor(
...
+}
+
+// TODO we can create another class for factories, it may be a copy-paste of this class or a subclass for some `AbstractEnumSerializer` with common `serialize`, `deserialize` and `toString` functions
 @PublishedApi
 @OptIn(ExperimentalSerializationApi::class)
 internal class EnumSerializer<T : Enum<T>>(
     serialName: String,
     private val values: Array<T>
 ) : KSerializer<T> {
+    @Volatile
```
Comment (sandwwraith) [inline core/commonMain/src/kotlinx/serialization/internal/Enums.kt:86]: Is `@Volatile` really needed here? I don't see why

Comment (shanshin) [reply core/commonMain/src/kotlinx/serialization/internal/Enums.kt:86]: Indeed, there is no need for it in the current implementation

**Review by sandwwraith — APPROVED** [review]

**Pull request merged**